libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFSDKCryptoUtils.m

```diff
@@ -230,6 +236,130 @@ + (nullable NSData*)decryptUsingRSAforData:(NSData *)data withKeyRef:(SecKeyRef)
+    
+    NSString *publicTagString = [NSString stringWithFormat:@"%@.%@", kSFECPublicKeyTagPrefix, keyName];
+    NSData *publicTag = [publicTagString dataUsingEncoding:NSUTF8StringEncoding];
+    
+    // Private key attributes
+    NSMutableDictionary* privateKeyAttributes =  [NSMutableDictionary new];
+    privateKeyAttributes[(id)kSecAttrIsPermanent] = @YES;
+    privateKeyAttributes[(id)kSecAttrApplicationTag] = privateTag;
+    
+    if (useSecureEnclave) {
+        CFErrorRef errorRef = NULL;
+        SecAccessControlRef privateAccess = SecAccessControlCreateWithFlags(kCFAllocatorDefault,
+                                                                            accessibleAttribute,
+                                                                            kSecAccessControlPrivateKeyUsage,
+                                                                            &errorRef);
+        if (errorRef == errSecSuccess) {
+            privateKeyAttributes[(id)kSecAttrAccessControl] = (__bridge id)privateAccess;
+        }
+    } else {
+        privateKeyAttributes[(id)kSecAttrAccessible] = (__bridge id)accessibleAttribute;
+    }
+
+    // Public key attributes
+    NSDictionary* publicKeyAttributes = @{ (id)kSecAttrIsPermanent:    @YES,
+                                           (id)kSecAttrApplicationTag: publicTag,
+                                           (id)kSecAttrAccessible: (__bridge id)accessibleAttribute,
+                                        };
+
+    // Key pair attributes
+    NSMutableDictionary* keyPairAttributes =  [NSMutableDictionary new];
+    keyPairAttributes[(id)kSecAttrKeyType] = (id)kSecAttrKeyTypeECSECPrimeRandom;
+    keyPairAttributes[(id)kSecAttrKeySizeInBits] = @256;
+    if (useSecureEnclave) {
+        keyPairAttributes[(id)kSecAttrTokenID] = (id)kSecAttrTokenIDSecureEnclave;
+    }
+    keyPairAttributes[(id)kSecPrivateKeyAttrs] = privateKeyAttributes;
+    keyPairAttributes[(id)kSecPublicKeyAttrs] = publicKeyAttributes;
+    
+    // Generation
+    CFErrorRef error = NULL;
+    SecKeyRef privateKey = SecKeyCreateRandomKey((__bridge CFDictionaryRef)keyPairAttributes, &error);
+    if (privateKey == nil) {
+        NSError *err = CFBridgingRelease(error);
+        // Handle the error. . .
+        [SFSDKCoreLogger e:[self class] format:@"Error creating EC private Key with name %@.  Error code: %@", keyName, err.localizedDescription];
+        return;
+    }
+    
+    SecKeyRef publicKey = SecKeyCopyPublicKey(privateKey);
+    
+    if (publicKey == nil) {
+        [SFSDKCoreLogger e:[self class] format:@"Error creating EC public key with name %@.", keyName];
+    }
+    
+    if (publicKey != nil)  {
+        CFRelease(publicKey);
+    }
+    if (privateKey != nil) {
+        CFRelease(privateKey);
+    }
+}
+
++ (void)deleteECKeyPairWithName:(NSString *)keyName
+{
+    [SFSDKCryptoUtils deleteKeyByTag:[NSString stringWithFormat:@"%@.%@", kSFECPublicKeyTagPrefix, keyName]];
```

Comment:
What would happen in the case that there is a failure to delete the private key and not the public or vice versa?

Reply:
It's unlikely since they have the same access controls.
When creating/retrieving a key, we see if we can get both otherwise we consider that the key doesn't exist - we then either create a new key or return nil (depending on whether you do a create or a retrieve). See the getKeyRefs method.